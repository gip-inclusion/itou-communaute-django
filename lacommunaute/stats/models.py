from django.db import models

from lacommunaute.forum.models import Forum
from lacommunaute.stats.enums import Period


class StatQuerySet(models.QuerySet):
    def current_month_datas(self):
        qs = Stat.objects.filter(period="month").order_by("-date")

        if qs.exists():
            return Stat.objects.filter(date=qs.first().date, period=Period.MONTH).values("name", "value", "date")

        return Stat.objects.none()


class Stat(models.Model):
    name = models.CharField(max_length=30, verbose_name="Nom")
    date = models.DateField(verbose_name="Date")
    value = models.IntegerField(verbose_name="Valeur")
    period = models.CharField(max_length=10, verbose_name="Période", choices=Period.choices)

    objects = models.Manager()

    class Meta:
        verbose_name = "Stat"
        verbose_name_plural = "Stats"
        ordering = ["date", "period", "name"]
        unique_together = ("name", "date", "period")

    def __str__(self):
        return f"{self.name} - {self.date} - {self.period}"

    objects = StatQuerySet().as_manager()


class ForumStat(models.Model):
    date = models.DateField(verbose_name="Date")
    period = models.CharField(max_length=10, verbose_name="Période", choices=Period.choices)
    forum = models.ForeignKey(Forum, on_delete=models.SET_NULL, verbose_name="Forum", null=True)
    visits = models.IntegerField(verbose_name="Visites", default=0)
    entry_visits = models.IntegerField(verbose_name="Visites entrantes", default=0)
    time_spent = models.IntegerField(verbose_name="Temps passé", default=0)

    objects = models.Manager()

    class Meta:
        verbose_name = "Stat de forum"
        verbose_name_plural = "Stats de forum"
        ordering = ["date", "period", "forum"]
        unique_together = ("date", "period", "forum")

    def __str__(self):
        return f"{self.date} - {self.period} - {self.forum}"
