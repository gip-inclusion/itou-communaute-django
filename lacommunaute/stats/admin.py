from dateutil.relativedelta import relativedelta
from django.contrib import admin

from lacommunaute.forum.models import Forum
from lacommunaute.stats.models import ForumStat, Stat


class ForumWithStatsFilter(admin.SimpleListFilter):
    title = "Forum"
    parameter_name = "forum"

    def lookups(self, request, model_admin):
        forums_with_stats = model_admin.model.objects.values_list("forum", flat=True)
        return Forum.objects.filter(pk__in=forums_with_stats).values_list("pk", "name")

    def queryset(self, request, queryset):
        if self.value():
            return queryset.filter(forum=self.value())
        return queryset


class BaseStatAdmin(admin.ModelAdmin):
    list_display = ("explicit_period",)
    list_filter = ("date", "period")

    @admin.display(ordering="date")
    def explicit_period(self, obj):
        if obj.period == "month":
            return f"{obj.date} au {obj.date + relativedelta(months=1) - relativedelta(days=1)}"
        if obj.period == "week":
            return f"{obj.date} au {obj.date + relativedelta(days=6)}"
        return f"{obj.date} (jour unique)"


@admin.register(Stat)
class StatAdmin(BaseStatAdmin):
    list_display = BaseStatAdmin.list_display + ("name", "value")
    list_filter = BaseStatAdmin.list_filter + ("name",)


@admin.register(ForumStat)
class ForumStatAdmin(BaseStatAdmin):
    list_display = BaseStatAdmin.list_display + ("forum", "visits", "entry_visits", "time_spent")
    list_filter = BaseStatAdmin.list_filter + (ForumWithStatsFilter,)
    raw_id_fields = ("forum",)
