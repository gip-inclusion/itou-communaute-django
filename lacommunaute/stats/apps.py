from django.apps import AppConfig


class ForumStatsAppConfig(AppConfig):
    label = "stats"
    name = "lacommunaute.stats"
    verbose_name = "Forum Stats"
    verbose_name_plural = "Forum Stats"
