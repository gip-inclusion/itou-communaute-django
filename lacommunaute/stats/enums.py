from django.db import models


class Period(models.TextChoices):
    MONTH = "month"
    WEEK = "week"
    DAY = "day"
