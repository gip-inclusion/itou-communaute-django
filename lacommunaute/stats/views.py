import logging

from dateutil.relativedelta import relativedelta
from django.db.models import Avg, CharField, Count, OuterRef, Subquery, Sum
from django.db.models.functions import Cast
from django.shortcuts import render
from django.utils.dateformat import format
from django.utils.timezone import localdate
from django.views import View
from django.views.generic.base import TemplateView

from lacommunaute.forum.models import Forum, ForumRating
from lacommunaute.stats.models import Stat
from lacommunaute.surveys.models import DSP
from lacommunaute.utils.json import extract_values_in_list
from lacommunaute.utils.math import percent


logger = logging.getLogger(__name__)


def get_daily_visits_stats(from_date, to_date):
    indicator_names = [
        "nb_uniq_visitors",
        "nb_uniq_engaged_visitors",
    ]
    datas = (
        Stat.objects.filter(period="day", name__in=indicator_names, date__gte=from_date, date__lte=to_date)
        .values("name", "value")
        .annotate(date=Cast("date", CharField()))
    )
    return extract_values_in_list(datas, indicator_names)


class StatistiquesPageView(TemplateView):
    template_name = "stats/statistiques.html"

    def get_funnel_data(self):
        qs = Stat.objects.current_month_datas()

        stats = {
            "period": None,
            "nb_uniq_visitors": 0,
            "nb_uniq_active_visitors": 0,
            "nb_uniq_engaged_visitors": 0,
        }

        if qs.filter(name="nb_uniq_visitors").exists():
            stats["period"] = format(qs.get(name="nb_uniq_visitors")["date"], "F Y")
            stats["nb_uniq_visitors"] = qs.get(name="nb_uniq_visitors")["value"]

        if qs.filter(name="nb_uniq_active_visitors").exists():
            stats["nb_uniq_active_visitors"] = qs.get(name="nb_uniq_active_visitors")["value"]

        if qs.filter(name="nb_uniq_engaged_visitors").exists():
            stats["nb_uniq_engaged_visitors"] = qs.get(name="nb_uniq_engaged_visitors")["value"]

        stats["activation_percent"] = percent(stats["nb_uniq_active_visitors"], stats["nb_uniq_visitors"])
        stats["engagement_percent"] = percent(stats["nb_uniq_engaged_visitors"], stats["nb_uniq_active_visitors"])
        return stats

    def get_monthly_visitors(self):
        indicator_names = ["nb_uniq_visitors_returning"]
        datas = (
            Stat.objects.filter(period="month", name__in=indicator_names)
            .values("name", "value")
            .annotate(date=Cast("date", CharField()))
        )
        return extract_values_in_list(datas, indicator_names)

    def get_dsp_count(self):
        return DSP.objects.count()

    def get_context_data(self, **kwargs):
        yesterday = localdate() - relativedelta(days=1)
        context = super().get_context_data(**kwargs)
        context["stats"] = get_daily_visits_stats(from_date=yesterday - relativedelta(days=89), to_date=yesterday)
        context["impact"] = self.get_monthly_visitors()
        context["dsp_count"] = self.get_dsp_count()
        context = {**context, **self.get_funnel_data()}

        return context


class BaseDetailStatsView(TemplateView):
    def get_detailled_stats(self):
        datas = (
            Stat.objects.filter(
                period=self.period,
                name__in=self.indicator_names,
                date__gt=localdate() - relativedelta(months=self.months),
            )
            .values("name", "value")
            .annotate(date=Cast("date", CharField()))
        )
        return extract_values_in_list(datas, self.indicator_names)

    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs)
        context["stats"] = self.get_detailled_stats()
        context["box_title"] = self.box_title

        return context


class MonthlyVisitorsView(BaseDetailStatsView):
    template_name = "stats/monthly_visitors.html"
    box_title = "Utilisateurs uniques mensuels"
    indicator_names = [
        "nb_uniq_visitors",
        "nb_uniq_active_visitors",
        "nb_uniq_engaged_visitors",
        "nb_uniq_visitors_returning",
    ]
    period = "month"
    months = 9


class DailyDSPView(BaseDetailStatsView):
    template_name = "stats/daily_dsp.html"
    box_title = "Diagnostics Parcours IAE quotidiens"
    indicator_names = ["dsp"]
    period = "day"
    months = 3


class DocumentStatsView(View):
    def get_objects_with_stats_and_ratings(self):
        objects = (
            Forum.objects.filter(parent__type=Forum.FORUM_CAT, forumstat__period="week")
            .annotate(sum_visits=Sum("forumstat__visits"))
            .annotate(sum_time_spent=Sum("forumstat__time_spent"))
            .select_related("parent", "partner")
            .order_by("id")
        )
        ratings = ForumRating.objects.filter(forum=OuterRef("pk")).values("forum")
        return objects.annotate(
            avg_rating=Subquery(ratings.annotate(avg_rating=Avg("rating")).values("avg_rating")),
            count_rating=Subquery(ratings.annotate(count_rating=Count("rating")).values("count_rating")),
        )

    def get_sort_fields(self):
        return [
            {"key": "parent", "label": "Cat√©gorie"},
            {"key": "sum_time_spent", "label": "Temps de lecture"},
            {"key": "sum_visits", "label": "Nombre de Visites"},
            {"key": "count_rating", "label": "Nombres de notations"},
            {"key": "avg_rating", "label": "Moyenne des notations"},
        ]

    def get(self, request, *args, **kwargs):
        objects = self.get_objects_with_stats_and_ratings()
        sort_key = (
            request.GET.get("sort")
            if request.GET.get("sort") in [field["key"] for field in self.get_sort_fields()]
            else "sum_time_spent"
        )
        objects = objects.order_by("-" + sort_key)

        return render(
            request,
            "stats/documents.html",
            {
                "objects": objects,
                "sort_key": sort_key,
                "sort_fields": self.get_sort_fields(),
            },
        )
