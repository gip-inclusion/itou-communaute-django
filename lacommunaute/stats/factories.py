import datetime

import factory
import factory.django

from lacommunaute.forum.factories import ForumFactory
from lacommunaute.stats.enums import Period
from lacommunaute.stats.models import ForumStat, Stat


class StatFactory(factory.django.DjangoModelFactory):
    date = factory.Faker("date")
    name = factory.Faker("word")
    period = Period.DAY
    value = factory.Faker("pyint")

    class Meta:
        model = Stat

    class Params:
        for_dsp_snapshot = factory.Trait(
            date=datetime.date(2024, 5, 17),
            name="dsp",
            value=46,
            period="day",
        )


class ForumStatFactory(factory.django.DjangoModelFactory):
    date = factory.Faker("date")
    period = Period.DAY
    forum = factory.SubFactory(ForumFactory)
    visits = factory.Faker("pyint")
    entry_visits = factory.Faker("pyint")
    time_spent = factory.Faker("pyint")

    class Meta:
        model = ForumStat

    class Params:
        for_snapshot = factory.Trait(period="week", date=datetime.date(2024, 5, 20))
        for_snapshot_older = factory.Trait(period="week", date=datetime.date(2024, 5, 13))
