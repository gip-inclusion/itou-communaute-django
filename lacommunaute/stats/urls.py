from django.urls import path

from lacommunaute.stats.views import DailyDSPView, ForumStatWeekArchiveView, MonthlyVisitorsView, StatistiquesPageView


app_name = "stats"

urlpatterns = [
    path("", StatistiquesPageView.as_view(), name="statistiques"),
    path("monthly-visitors/", MonthlyVisitorsView.as_view(), name="monthly_visitors"),
    path("dsp/", DailyDSPView.as_view(), name="dsp"),
    path("weekly/<int:year>/<int:week>/", ForumStatWeekArchiveView.as_view(), name="forum_stat_week_archive"),
]
