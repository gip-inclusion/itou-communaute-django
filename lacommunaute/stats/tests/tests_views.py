from datetime import date

import pytest  # noqa
from dateutil.relativedelta import relativedelta
from django.test import TestCase
from django.urls import reverse
from django.utils import timezone
from django.utils.dateformat import format
from django.utils.timezone import localdate
from faker import Faker
from freezegun import freeze_time
from machina.core.loading import get_class
from pytest_django.asserts import assertContains, assertNotContains

from lacommunaute.documentation.factories import CategoryFactory, DocumentFactory, DocumentRatingFactory
from lacommunaute.stats.enums import Period
from lacommunaute.stats.factories import ForumStatFactory, StatFactory
from lacommunaute.surveys.factories import DSPFactory
from lacommunaute.utils.math import percent
from lacommunaute.utils.testing import parse_response_to_soup


faker = Faker()
assign_perm = get_class("forum_permission.shortcuts", "assign_perm")


class StatistiquesPageTest(TestCase):
    def test_context_data(self):
        url = reverse("stats:statistiques")
        response = self.client.get(url)
        self.assertEqual(response.status_code, 200)
        self.assertTemplateUsed(response, "stats/statistiques.html")

    def test_month_datas_in_context(self):
        today = localdate()
        url = reverse("stats:statistiques")

        # no data
        response = self.client.get(url)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.context["period"], None)
        self.assertEqual(response.context["nb_uniq_visitors"], 0)
        self.assertEqual(response.context["nb_uniq_active_visitors"], 0)
        self.assertEqual(response.context["nb_uniq_engaged_visitors"], 0)
        self.assertEqual(response.context["activation_percent"], 0)
        self.assertEqual(response.context["engagement_percent"], 0)

        # undesired data
        StatFactory(name="nb_uniq_engaged_visitors", period=Period.DAY, date=today)
        StatFactory(name=faker.word(), period=Period.MONTH, date=today)
        response = self.client.get(url)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.context["period"], None)
        self.assertEqual(response.context["nb_uniq_visitors"], 0)
        self.assertEqual(response.context["nb_uniq_active_visitors"], 0)
        self.assertEqual(response.context["nb_uniq_engaged_visitors"], 0)
        self.assertEqual(response.context["activation_percent"], 0)
        self.assertEqual(response.context["engagement_percent"], 0)

        uniq_visitors = StatFactory(name="nb_uniq_visitors", period=Period.MONTH, date=today)
        uniq_active_visitors = StatFactory(name="nb_uniq_active_visitors", period=Period.MONTH, date=today)
        uniq_engaged_visitors = StatFactory(name="nb_uniq_engaged_visitors", period=Period.MONTH, date=today)
        response = self.client.get(url)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.context["period"], format(today, "F Y"))
        self.assertEqual(response.context["nb_uniq_visitors"], uniq_visitors.value)
        self.assertEqual(response.context["nb_uniq_active_visitors"], uniq_active_visitors.value)
        self.assertEqual(response.context["nb_uniq_engaged_visitors"], uniq_engaged_visitors.value)
        self.assertEqual(
            response.context["activation_percent"], percent(uniq_active_visitors.value, uniq_visitors.value)
        )
        self.assertEqual(
            response.context["engagement_percent"], percent(uniq_engaged_visitors.value, uniq_active_visitors.value)
        )

    def test_impact_in_context_data(self):
        url = reverse("stats:statistiques")
        today = localdate()
        empty_res = {"date": [], "nb_uniq_visitors_returning": []}

        # no data
        response = self.client.get(url)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.context["impact"], empty_res)

        # undesired data
        StatFactory(name="nb_uniq_visitors_returning", period=Period.DAY, date=today)
        StatFactory(name=faker.word(), period=Period.MONTH, date=today)
        response = self.client.get(url)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.context["impact"], empty_res)

        # desired data
        StatFactory(name="nb_uniq_visitors_returning", period=Period.MONTH, date=today, value=1)
        response = self.client.get(url)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.context["impact"]["date"][0], today.strftime("%Y-%m-%d"))
        self.assertEqual(response.context["impact"]["nb_uniq_visitors_returning"][0], 1)

    def test_navigation(self):
        url = reverse("stats:statistiques")
        response = self.client.get(url)
        self.assertContains(response, f"<a href={reverse('stats:monthly_visitors')}>")


@pytest.fixture(name="setup_statistiques_data")
def setup_statistiques_data_fixture(request):
    last_visible_date = date(2024, 6, 30)
    first_visible_date = last_visible_date - timezone.timedelta(days=89)
    if request.param == "undesired_data_setup":
        return [
            StatFactory(name="nb_uniq_visitors", date=first_visible_date, period=Period.MONTH),
            StatFactory(name="nb_uniq_visitors", date=first_visible_date, period=Period.WEEK),
            StatFactory(name=faker.word(), date=first_visible_date, period=Period.DAY),
            StatFactory(
                name="nb_uniq_visitors", date=first_visible_date - timezone.timedelta(days=1), period=Period.DAY
            ),
            StatFactory(
                name="nb_uniq_visitors", date=last_visible_date + timezone.timedelta(days=1), period=Period.DAY
            ),
        ]
    if request.param == "desired_data_setup":
        stats = [
            ("nb_uniq_visitors", first_visible_date, 8469),
            ("nb_uniq_visitors", last_visible_date, 8506),
            ("nb_uniq_engaged_visitors", first_visible_date, 128),
            ("nb_uniq_engaged_visitors", last_visible_date, 5040),
        ]
        return [StatFactory(name=name, date=date, value=value) for name, date, value in stats]
    return None


class TestStatistiquesPageView:
    def test_dsp_count(self, client, db, snapshot):
        DSPFactory.create_batch(10)
        url = reverse("stats:statistiques")
        response = client.get(url)
        assert response.status_code == 200
        assert str(parse_response_to_soup(response, selector="#daily_dsp")) == snapshot(name="dsp")

    @pytest.mark.parametrize(
        "setup_statistiques_data,expected",
        [
            (
                None,
                {"date": [], "nb_uniq_visitors": [], "nb_uniq_engaged_visitors": []},
            ),
            (
                "undesired_data_setup",
                {"date": [], "nb_uniq_visitors": [], "nb_uniq_engaged_visitors": []},
            ),
            (
                "desired_data_setup",
                {
                    "date": ["2024-04-02", "2024-06-30"],
                    "nb_uniq_visitors": [8469, 8506],
                    "nb_uniq_engaged_visitors": [128, 5040],
                },
            ),
        ],
        indirect=["setup_statistiques_data"],
    )
    def test_visitors_in_context_data(self, client, db, setup_statistiques_data, expected):
        with freeze_time("2024-07-01"):
            url = reverse("stats:statistiques")
            response = client.get(url)
        assert response.status_code == 200
        assert response.context["stats"] == expected

    def test_link_to_document_stats_view(self, client, db):
        url = reverse("stats:statistiques")
        response = client.get(url)
        assert response.status_code == 200
        assertContains(response, reverse("stats:document_stats"))


class TestMonthlyVisitorsView:
    def test_context_data(self, client, db):
        url = reverse("stats:monthly_visitors")
        today = localdate()
        empty_res = {
            "date": [],
            "nb_uniq_visitors": [],
            "nb_uniq_active_visitors": [],
            "nb_uniq_engaged_visitors": [],
            "nb_uniq_visitors_returning": [],
        }

        # no data
        response = client.get(url)
        assert response.status_code == 200
        assertContains(response, "Utilisateurs uniques mensuels")
        assert response.context["box_title"] == "Utilisateurs uniques mensuels"
        assert response.context["stats"] == empty_res

        # undesired data
        StatFactory(name="nb_uniq_visitors_returning", period=Period.DAY, date=today)
        StatFactory(name=faker.word(), period=Period.MONTH, date=today)
        StatFactory(name="nb_uniq_visitors", period=Period.MONTH, date=today - relativedelta(months=9), value=1)
        response = client.get(url)
        assert response.status_code == 200
        assert response.context["stats"] == empty_res

        # expected data
        StatFactory(name="nb_uniq_visitors_returning", period=Period.MONTH, date=today, value=2)
        StatFactory(name="nb_uniq_visitors", period=Period.MONTH, date=today - relativedelta(months=8), value=10)
        response = client.get(url)
        assert response.status_code == 200
        assert response.context["stats"] == {
            "date": [(today - relativedelta(months=8)).strftime("%Y-%m-%d"), today.strftime("%Y-%m-%d")],
            "nb_uniq_visitors": [10],
            "nb_uniq_active_visitors": [],
            "nb_uniq_engaged_visitors": [],
            "nb_uniq_visitors_returning": [2],
        }

    def test_navigation(self, client, db, snapshot):
        url = reverse("stats:dsp")
        response = client.get(url)
        assert response.status_code == 200
        assert str(parse_response_to_soup(response, selector=".c-breadcrumb")) == snapshot(name="breadcrumb")


class TestDailyDSPView:
    def test_context_data(self, client, db):
        today = localdate()

        # undesired datas
        StatFactory(name="dsp", period=Period.DAY, date=today - relativedelta(months=3), value=20)
        StatFactory(name=faker.word(), period=Period.DAY, date=today, value=21)

        # expected datas
        StatFactory(
            name="dsp", period=Period.DAY, date=today - relativedelta(months=3) + relativedelta(days=1), value=3
        )
        StatFactory(name="dsp", period=Period.DAY, date=today, value=2)

        url = reverse("stats:dsp")
        response = client.get(url)
        assert response.status_code == 200
        assert response.context["box_title"] == "Diagnostics Parcours IAE quotidiens"
        assert response.context["stats"] == {
            "date": [
                (today - relativedelta(months=3) + relativedelta(days=1)).strftime("%Y-%m-%d"),
                today.strftime("%Y-%m-%d"),
            ],
            "dsp": [3, 2],
        }

    def test_navigation(self, client, db, snapshot):
        url = reverse("stats:dsp")
        response = client.get(url)
        assert response.status_code == 200
        assert str(parse_response_to_soup(response, selector=".c-breadcrumb")) == snapshot(name="breadcrumb")


@pytest.fixture(name="document_stats_setup")
def document_stats_setup_fixture(db):
    category = CategoryFactory()
    fa = DocumentFactory(name="A", category=category)
    fb = DocumentFactory(name="B", category=category)
    fc = DocumentFactory(name="C", category=category)
    fd = DocumentFactory(name="D", category=category)
    # ForumStatFactory(forum=fa, period="week", visits=70, time_spent=40 * 60)
    # ForumStatFactory(forum=fb, period="week", visits=100, time_spent=30 * 60)
    # ForumStatFactory(forum=fc, period="week", visits=90, time_spent=20 * 60)
    # ForumStatFactory(forum=fd, period="week", visits=80, time_spent=10 * 60)
    DocumentRatingFactory.create_batch(2, document=fa, rating=4)
    DocumentRatingFactory.create_batch(1, document=fb, rating=3)
    DocumentRatingFactory.create_batch(4, document=fc, rating=2)
    DocumentRatingFactory.create_batch(3, document=fd, rating=5)

    # undesired forum
    DocumentFactory(name="Orphelin Document")
    DocumentFactory(name="Document wo DocumentStats", category=category)

    return category


class TestForumStatView:
    @pytest.mark.parametrize(
        "sort_key,snapshot_name",
        [
            (None, "sort_by_sum_time_spent"),
            ("sum_time_spent", "sort_by_sum_time_spent"),
            ("sum_visits", "sort_by_sum_visits"),
            ("avg_rating", "sort_by_avg_rating"),
            ("count_rating", "sort_by_count_rating"),
            ("unknown", "sort_by_sum_time_spent"),
        ],
    )
    def test_sort_key(self, client, db, document_stats_setup, sort_key, snapshot_name, snapshot):
        url = reverse("stats:document_stats") + f"?sort={sort_key}" if sort_key else reverse("stats:document_stats")
        response = client.get(url)
        assert response.status_code == 200
        assert str(
            parse_response_to_soup(
                response,
                selector="main",
                replace_in_href=[document for document in document_stats_setup.documents.all()],
            )
        ) == snapshot(name=snapshot_name)

    def test_num_queries(self, client, db, document_stats_setup, django_assert_num_queries):
        django_session_num_of_queries = 6
        expected_queries_in_view = 1
        with django_assert_num_queries(expected_queries_in_view + django_session_num_of_queries):
            client.get(reverse("stats:document_stats"))


class TestForumStatWeekArchiveView:
    def get_url_from_date(self, date):
        return reverse(
            "stats:forum_stat_week_archive", kwargs={"year": date.strftime("%Y"), "week": date.strftime("%W")}
        )

    def test_header_and_breadcrumb(self, client, db, snapshot):
        response = client.get(self.get_url_from_date(ForumStatFactory(for_snapshot=True).date))
        assert response.status_code == 200
        assert str(parse_response_to_soup(response, selector=".s-title-01")) == snapshot(name="title-01")
        assert str(parse_response_to_soup(response, selector=".c-breadcrumb")) == snapshot(name="breadcrumb")

    def test_navigation(self, client, db):
        weeks = [date.today() - relativedelta(weeks=i) for i in range(15, 10, -1)]
        for week in weeks[1:4]:
            ForumStatFactory(date=week, for_snapshot=True)

        test_cases = [
            {"test_week": weeks[1], "not_contains": [weeks[0]], "contains": [weeks[2]]},
            {"test_week": weeks[2], "not_contains": [], "contains": [weeks[1], weeks[3]]},
            {"test_week": weeks[3], "not_contains": [weeks[4]], "contains": [weeks[2]]},
        ]

        for test_case in test_cases:
            response = client.get(self.get_url_from_date(test_case["test_week"]))
            for week in test_case["not_contains"]:
                assertNotContains(response, self.get_url_from_date(week))
            for week in test_case["contains"]:
                assertContains(response, self.get_url_from_date(week))

        # out of bound
        for week in [weeks[0], weeks[4]]:
            response = client.get(self.get_url_from_date(week))
            assert response.status_code == 404

    def test_most_viewed_forums(self, client, db, snapshot):
        forums_stats = [
            ForumStatFactory(for_snapshot=True, visits=10, entry_visits=8, time_spent=1000, forum__name="Forum A"),
            ForumStatFactory(for_snapshot=True, visits=17, entry_visits=5, time_spent=1978, forum__name="Forum B"),
        ]

        response = client.get(self.get_url_from_date(forums_stats[0].date))
        assert response.status_code == 200
        assert str(
            parse_response_to_soup(
                response, selector="#most_viewed", replace_in_href=[fs.forum for fs in forums_stats]
            )
        ) == snapshot(name="most_viewed_forums")

    def test_paginated_most_viewed_forums(self, client, db):
        ForumStatFactory.create_batch(16, for_snapshot=True)
        response = client.get(reverse("stats:forum_stat_week_archive", kwargs={"year": 2024, "week": 21}))
        assert response.status_code == 200
        assert len(response.context_data["forum_stats"]) == 15

    def test_most_rated_documents(self, client, db, snapshot):
        # required to unblock the view, until ForumStat is migrated to the new model
        fs = ForumStatFactory(for_snapshot=True, forum__name="Forum A")
        document = DocumentFactory(name="Document A")

        # rating within range
        DocumentRatingFactory.create_batch(2, rating=5, document=document, set_created=fs.date)
        # rating out of range
        DocumentRatingFactory.create_batch(2, rating=1, document=document)

        # undesired document
        DocumentFactory()

        # undesired rating
        DocumentRatingFactory(rating=4)

        response = client.get(self.get_url_from_date(fs.date))
        assert response.status_code == 200
        assert str(parse_response_to_soup(response, selector="#most_rated")) == snapshot(name="most_rated_documents")

    def test_visitors(self, client, db, snapshot):
        fs = ForumStatFactory(for_snapshot=True)

        # relevant
        relevant_dates = [
            fs.date,
            fs.date + relativedelta(days=6),
            fs.date + relativedelta(days=6) - relativedelta(days=89),
        ]
        visitors = [10, 11, 12]
        for stat_date, visitor_count in zip(relevant_dates, visitors):
            StatFactory(date=stat_date, name="nb_uniq_visitors", value=visitor_count)

        # undesired
        for stat_date in [fs.date + relativedelta(weeks=1), fs.date + relativedelta(days=6) - relativedelta(days=90)]:
            StatFactory(date=stat_date, name="nb_uniq_visitors", value=99)

        response = client.get(self.get_url_from_date(fs.date))
        assert response.status_code == 200
        expected_stats = {
            "date": ["2024-02-27", "2024-05-20", "2024-05-26"],
            "nb_uniq_visitors": [12, 10, 11],
            "nb_uniq_engaged_visitors": [],
        }
        assert response.context_data["stats"] == expected_stats


@pytest.mark.parametrize(
    "forum_stats,status_code",
    [
        (lambda: None, 404),
        (lambda: [ForumStatFactory(for_snapshot=True), ForumStatFactory(for_snapshot_older=True)], 302),
    ],
)
def test_redirect_to_latest_weekly_stats(client, db, forum_stats, status_code):
    forum_stats = forum_stats()
    response = client.get(reverse("stats:redirect_to_latest_weekly_stats"))
    assert response.status_code == status_code
    if forum_stats:
        assert response.url == reverse(
            "stats:forum_stat_week_archive",
            kwargs={"year": forum_stats[0].date.year, "week": forum_stats[0].date.strftime("%W")},
        )
