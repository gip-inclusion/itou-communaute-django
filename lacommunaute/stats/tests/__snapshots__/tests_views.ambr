# serializer version: 1
# name: TestDailyDSPView.test_navigation[breadcrumb]
  '''
  <nav aria-label="Fil d'ariane" class="c-breadcrumb">
              <ol class="breadcrumb">
                  <li class="breadcrumb-item">Retourner vers</li>
                  <li class="breadcrumb-item">
                      <a href="/statistiques/">Statistiques</a>
                  </li>
              </ol>
          </nav>
  '''
# ---
# name: TestForumStatWeekArchiveView.test_header_and_breadcrumb[breadcrumb]
  '''
  <nav aria-label="Fil d'ariane" class="c-breadcrumb">
              <ol class="breadcrumb">
                  <li class="breadcrumb-item">Retourner vers</li>
                  <li class="breadcrumb-item">
                      <a href="/statistiques/">Statistiques</a>
                  </li>
              </ol>
          </nav>
  '''
# ---
# name: TestForumStatWeekArchiveView.test_header_and_breadcrumb[title-01]
  '''
  <section class="s-title-01 mt-lg-5">
          <div class="s-title-01__container container">
              <div class="s-title-01__row row">
                  <div class="s-title-01__col col-12">
                      <h1 class="s-title-01__title h1">
                          <strong>Partenariat Academie France Travail x La communauté de l'inclusion</strong>
                          <br/>
                          Statistiques de la semaine du 20 mai 2024 au 26 mai 2024
                      </h1>
                  </div>
              </div>
          </div>
      </section>
  '''
# ---
# name: TestForumStatWeekArchiveView.test_most_rated_forums[most_rated_forums]
  '''
  <div class="s-section__row row" id="most_rated">
                  <div class="s-section__col col-12">
                      <div class="c-box mb-3 mb-md-5">
                          <h2>Les 1 fiches pratiques les plus notées sur la période</h2>
                          <table class="table">
                              <thead>
                                  <tr>
                                      <th scope="col">Fiche Pratique</th>
                                      <th scope="col">Nouvelles otations sur la période</th>
                                      <th scope="col">Evaluation globale</th>
                                  </tr>
                              </thead>
                              <tbody>
                                  
                                      <tr>
                                          <th scope="row">Forum A</th>
                                          <td>2</td>
                                          <td>3,00</td>
                                      </tr>
                                  
                              </tbody>
                          </table>
                      </div>
                  </div>
              </div>
  '''
# ---
# name: TestForumStatWeekArchiveView.test_most_viewed_forums[most_viewed_forums]
  '''
  <div class="s-section__row row" id="most_viewed">
                  <div class="s-section__col col-12">
                      <div class="c-box mb-3 mb-md-5">
                          <h2>Les 2 fiches pratiques les plus lues sur la période</h2>
                          <table class="table">
                              <thead>
                                  <tr>
                                      <th scope="col">Fiche Pratique</th>
                                      <th scope="col">Visiteurs uniques</th>
                                      <th scope="col">Visiteurs uniques entrants</th>
                                      <th scope="col">Temps de lecture total</th>
                                  </tr>
                              </thead>
                              <tbody>
                                  
                                      <tr>
                                          <th scope="row">Forum B</th>
                                          <td>17</td>
                                          <td>5</td>
                                          <td>1978 secondes</td>
                                      </tr>
                                  
                                      <tr>
                                          <th scope="row">Forum A</th>
                                          <td>10</td>
                                          <td>8</td>
                                          <td>1000 secondes</td>
                                      </tr>
                                  
                              </tbody>
                          </table>
                      </div>
                  </div>
              </div>
  '''
# ---
# name: TestMonthlyVisitorsView.test_navigation[breadcrumb]
  '''
  <nav aria-label="Fil d'ariane" class="c-breadcrumb">
              <ol class="breadcrumb">
                  <li class="breadcrumb-item">Retourner vers</li>
                  <li class="breadcrumb-item">
                      <a href="/statistiques/">Statistiques</a>
                  </li>
              </ol>
          </nav>
  '''
# ---
# name: TestStatistiquesPageView.test_dsp_count[dsp]
  '''
  <div class="s-section__row row" id="daily_dsp">
                  <div class="s-section__col col-12">
                      <div class="c-box mb-3 mb-md-5">
                          <h2>10 Diagnostiques Parcours IAE</h2>
                          <p>
                              Réalisés sur la communauté de l'inclusion depuis le 24 février 2024
                              <span class="fs-sm">(<a href="/statistiques/dsp/">Voir les détails</a>)</span>
                          </p>
                      </div>
                  </div>
              </div>
  '''
# ---
